src/clj/fluree/db/reasoner/core.cljc

```diff
@@ -0,0 +1,151 @@
+      (log/debug "reasoner flakes: " rule-id reasoner-flakes)
+      ;; returns map of :db-after, :add, :remove - but for reasoning we only support adds, so remove should be empty
+      (transact/final-db db reasoner-flakes tx-state*))))
+
+(defn execute-reasoner
+  "Executes the reasoner on the staged db-after and returns the updated db-after."
+  [db reasoning-rules fuel-tracker reasoner-max tx-state]
+  (go-try
+    (let [rule-schedule (schedule reasoning-rules)]
+      (log/debug "reasoning schedule: " rule-schedule)
+      (if (seq rule-schedule)
+        (loop [[rule-id & r] rule-schedule
+               reasoned-flakes nil ;; Note these are in an AVL set in with spot comparator
+               reasoned-db     db
+               summary         {:iterations      0 ;; holds summary across all runs
+                                :reasoned-flakes []
+                                :total-flakes    0}]
+          (if rule-id
+            (let [{:keys [db-after add]} (<? (execute-reasoner-rule reasoned-db rule-id reasoning-rules fuel-tracker tx-state))]
+              (log/debug "executed reasoning rule: " rule-id)
+              (log/trace "reasoning rule: " rule-id "produced flakes:" add)
+              (recur r
+                     (if reasoned-flakes
+                       (into reasoned-flakes add)
+                       add)
+                     db-after
+                     summary))
+            (let [all-reasoned-flakes (into reasoned-flakes (:reasoned-flakes summary))
+                  summary*            {:iterations      (-> summary :iterations inc)
+                                       :reasoned-flakes all-reasoned-flakes
+                                       :total-flakes    (count all-reasoned-flakes)}
+                  new-flakes?         (> (:total-flakes summary*)
+                                         (:total-flakes summary))
+                  maxed?              (when reasoner-max
+                                        (= (:iterations summary*) reasoner-max))]
+
+              (log/debug "Total reasoned flakes:" (:total-flakes summary*))
+              "completed in:" (:iterations summary*) "iteration(s)."
+
+              (if (and new-flakes? (not maxed?))
+                (recur rule-schedule nil reasoned-db summary*)
+                (do
+                  (when (and maxed? new-flakes?)
+                    (log/warn (str "Reasoner reached max iterations: " reasoner-max
+                                   ". Returning db reasoned thus far.")))
+                  reasoned-db)))))
+        db))))
+
+(defn rules-from-graph
+  [graph]
+  (let [expanded (-> graph
+                     json-ld/expand
+                     util/sequential)]
+    (reduce
+      (fn [acc rule]
+        (let [id   (:id rule)
+              rule (util/get-first-value rule "http://flur.ee/ns/ledger#rule")]
+          (if rule
+            (conj acc [(or id (str "_:" (rand-int 2147483647))) rule])
```

Comment:
Well, maybe it could just be random and not secure random... but no judgement on paranoia. :)

Addressed in commit 655964edbbd09088dec664cb615d18e0d7c73bda